### PR TITLE
Code cleanup, expand README, fix some minor bugs

### PR DESCRIPTION
**Pull request by 3vorp**
(from changelog, make sure to add the release number and date when this gets merged)

### Added

- Exposed `Collection.collectionName` as a readonly property for TypeScript usage.
- TypeScript overview to the README.
- Optional replacement argument for `array-splice` edit fields.
- `array-contains-none` option for array fields.
- Optional constructor for the `JSONDatabase` PHP class to reduce repetitive code.
- "Advanced" section to the README for previously undocumented features.
- `original` option for `readRaw` to not insert ID fields, for easier non-relational collection usage.

### Changed

- Rejected incorrect parameters are now `TypeError`s instead of regular `Error`s.
- Deprecated `firestorm.table(name)` method, since `firestorm.collection(name)` does exactly the same thing.
- Reformatted the repository and improved README.md to make it easier to set up Firestorm.
- Clean up and standardize JSDoc comments.
- `editField` and `editFieldBulk` now return confirmations like all other write methods.
- `editField` and `editFieldBulk` now reject with a descriptive error message on failure rather than silently failing.

### Fixed

- PHP-level errors not being rejected properly in GET requests.
- Certain write commands mutating data internally and affecting parameters outside Firestorm.
- `Collection.searchKeys` and `Collection.values` not returning proper `Error` objects sometimes.
- `files.upload` not allowing the `form-data` package's typing of `FormData` in TypeScript.
- Inconsistent use of indentation and formatting in PHP files.
- Various typos in PHP files.
- `Collection` class being exported in TypeScript despite the actual class being private.
- `array-splice` edit fields being incorrectly typed as `array-slice`.
- Platform-specific PHP error when searching nested keys.
- `Collection.remove` rejecting numeric keys, despite `Collection.removeBulk` not doing so.
- `editField` and `editFieldBulk` validation issues.

### CODE REVIEW

File: tests/js-test.spec.js [hunk 1]
```diff
@@ -274,18 +273,15 @@ describe("GET operations", () => {
 				.then(() => {
 					done(new Error("Parameter should be an array of string or number"));
 				})
-				.catch((err) => {
-					expect(err).to.equal("Incorrect keys");
-					done();
```
Comment (TheRolfFR) [inline tests/js-test.spec.js:279]: why did you delete error assert?

File: typings/index.d.ts [hunk 2]
```diff
@@ -205,7 +208,7 @@ export class Collection<T> {
 	 * @returns The found elements
 	 */
 	public search(
-		options: SearchOption<RemoveMethods<T> & { id: string | number }>[],
+		options: SearchOption<RemoveMethods<T> & { id: string }>[],
```
Comment (TheRolfFR) [inline typings/index.d.ts:211]: it might create errors if some old functions still use a number

File: typings/index.d.ts [hunk 3]
```diff
@@ -237,7 +240,7 @@ export class Collection<T> {
 	 */
 	public select<K extends Array<"id" | keyof T>>(
 		option: SelectOption<K>,
-	): Promise<Record<string, Pick<T & { id: string | number }, K[number]>>>;
+	): Promise<Record<string, Pick<T & { id: string }, K[number]>>>;
```
Comment (TheRolfFR) [inline typings/index.d.ts:243]: same here

File: typings/index.d.ts [hunk 4]
```diff
@@ -356,13 +359,6 @@ export function token(value?: string): string;
  */
 export function collection<T>(value: string, addMethods?: CollectionMethods<T>): Collection<T>;
 
-/**
- * Create a temporary Firestorm collection with no methods
- * @param table - The table name to get
- * @returns The collection
- */
-export function table<T>(table: string): Promise<Collection<T>>;
-
```
Comment (TheRolfFR) [inline typings/index.d.ts:365]: or you could set the method as deprecated

**Review by TheRolfFR — CHANGES_REQUESTED** [review]

File: tests/js-test.spec.js (hunk 1 above)
Comment (3vorp) [reply tests/js-test.spec.js:279]: the error message was previously inconsistent with all the others, this is more similar to what other methods do

File: typings/index.d.ts (hunk 4 above)
Comment (3vorp) [reply typings/index.d.ts:365]: smart, will do

File: typings/index.d.ts (hunk 2 above)
Comment (TheRolfFR) [reply typings/index.d.ts:217]: Keep it please

File: tests/config.php [hunk 5]
```diff
@@ -15,18 +15,18 @@
...
+// test with constructor/optional args
+$database_list['house'] = new JSONDatabase('house', false);
+
+// test without constructor
+$tmp = new JSONDatabase;
+$tmp->fileName = 'base';
+$tmp->autoKey = true;
+$tmp->autoIncrement = true;
+
+$database_list[$tmp->fileName] = $tmp;
```
Comment (TheRolfFR) [inline tests/config.php:28]: I suppose you can delete $props

**Review by TheRolfFR — CHANGES_REQUESTED** [review]

**Review by TheRolfFR — APPROVED** [review]

File: tests/tokens.php [hunk 6]
```diff
@@ -1,7 +1,3 @@
 <?php
 
-$db_tokens_map = array(
-    'Test' => 'NeverGonnaGiveYouUp'
-);
-
-$db_tokens = array_values($db_tokens_map);
\ No newline at end of file
+$db_tokens = array("NeverGonnaGiveYouUp");
```
Comment (TheRolfFR) [inline tests/tokens.php]: I don't like that, I keep maps so that I can identify which token is used where and better know what tokens to revoke.

On top of being useless, this modification destroys meaning

**Review by TheRolfFR — CHANGES_REQUESTED** [review]
For real, what's the point of modifying my code, you're just modifying code to modify code.

It's useless, and you removed my ``sec`` function. Why, I don't know, but if I made it, it's so that I can modify the content later if needed

**Comment by 3vorp** [comment]
I'm just trying to make the code cleaner, it was really inconsistent before and most of the changes are simply running a formatter and linter.

**Review by TheRolfFR — APPROVED** [review]

**Review by TheRolfFR — APPROVED** [review]
